// egoomongodb/index.js
var MongoClient = require('mongodb').MongoClient,
    ejs = require('ejs'),
    fs = require('fs'),
    uuid = require('node-uuid'),
    mongodbURL = require('../../package.json').mongodburl;

//console.log();
var connectAndExc= function (callback) {
    MongoClient.connect(mongodbURL, function(err, db) {
        var collectionUser = db.collection('user');
        callback(collectionUser,db);
    });
};

function EgooMongoClient(){

}

EgooMongoClient.prototype.validLogin = validLogin;
//登陆验证
var validLogin = function (userInfo, res) {
    //console.log(userInfo);
    connectAndExc(function (collectionUser,db) {
        collectionUser.find(userInfo).toArray(function (err,items) {
            if(err) throw err;
            //console.log(items);
            if(items.length === 0){
                res.send(JSON.stringify({result:false}));
            }else{
                indexejs = fs.readFileSync('./views/egoosocket.html', 'utf8');
                var index = ejs.render(indexejs,{"user":items[0]});
                res.writeHead(200, {'Content-type' : 'text/html'});
                res.write(index);
                res.end();
            }
            db.close();
        });
    });
};
/*
* 对应临时socketid
* userinfo
* {
*   "userid":"",
*   "tempsocketid":""
* }
* */
var mapTempSocketid = function (userinfo,res) {
    connectAndExc(function (collectionUser,db) {
        collectionUser.updateOne({"userid":userinfo.userid},{$set:{"tempsocketid":userinfo.tempsocketid}}, function (err,result) {
            if(err) throw err;
            db.close();
            var _result = {"mapTempResult":false};
            if(result.result.nModified > 0){
                _result.mapTempResult = true;
            }
            if(res){
                res.send(JSON.stringify(_result));
            }
        });
    });
};
/*
 * 清除对应socketid
 * { value:
 { _id: 561e144eb564ee9708fb28e6,
 userid: 'admin',
 username: 'admin',
 password: '1',
 tempsocketid: '' },
 lastErrorObject: { updatedExisting: true, n: 1 },
 ok: 1 }
 * */
var clearTempSocketid = function (socketid) {
    connectAndExc(function (collectionUser,db) {
        collectionUser.findAndModify({"tempsocketid":socketid},[],{$set:{"tempsocketid":""}},{new:true}, function (err,result) {
            if(err) throw err;
            db.close();
            //console.log(result);
            if(!result.value){
                console.log('清除socketid失败!');
            }
        });
    });
};

/*
*获取用户userid
* */
var getUseridBySocketid = function (socketid,res) {
    connectAndExc(function (collectionUser,db) {
        collectionUser.find(socketid).toArray(function (err,result) {
            if(err) throw err;
            //console.log(result);
            if(result.length !== 0){
                db.close();
                res.send(JSON.stringify({"username":result[0].username,"userid":result[0].userid,"tempsocketid":socketid.tempsocketid}));
            }else{
                setTimeout(function () {
                    collectionUser.find(socketid).toArray(function (err,result) {
                        if(err) throw err;
                        db.close();
                        if(result.length !== 0){
                            res.send(JSON.stringify({"username":result[0].username,"userid":result[0].userid,"tempsocketid":socketid.tempsocketid}));
                        }else{
                            res.send(JSON.stringify({"username":"","userid":"","tempsocketid":socketid.tempsocketid}));
                        }
                    });
                },1000);
            }
        });
    });
};

/*
*incomingdata
* {
*   inconfirm:true/false,
*   userid:"",
*   touserid : "",
*   msguuid:"",
*   msgdatatime:"",
*   msgcontent:"",
*
*
* }
* */
var recordSendMsg = function (incomingdata,res) {
    //查询是否存在该聊天纪录
    connectAndExc(function (collections, db) {
        var query = {
            "userid":incomingdata.userid
        };
        var msgsString = '{ "msgs.'+incomingdata.touserid+'" : true }';
        var suppression = JSON.parse(msgsString);
        collections.find(query,suppression).toArray(function (err,result) {
            if(err) throw err;
            //console.log();
            //[ { _id: 561e144eb564ee9708fb28e7, msgs: { gengyuchen: '1' } } ]

            //如果不存在该聊天纪录,创建
            if(result[0] && isEmptyObject(result[0].msgs)){
                var selector = {"userid":incomingdata.userid};
                var _uuid = uuid.v4();
                var msgobj = '{"msgs.'+incomingdata.touserid+'" : "'+_uuid+'"}';
                //console.log(msgobj);
                var document = {$set : JSON.parse(msgobj)};
                collections.update(selector,document,{'w':1}, function (err, result) {
                    if(err) throw err;
                    var msgscollections = db.collection('msg');
                    var insertmsguuid = JSON.parse('{"'+_uuid+'" : []}');
                    msgscollections.insertOne(insertmsguuid, {w:1}, function (err,_result) {
                        if(err) throw err;
                        //console.log('插入后');
                        //console.log(_result);
                    });
                });
            }else{
                var msgsuuid = result[0].msgs[incomingdata.touserid];
                console.log('存在');
                console.log(msgsuuid);
            }
        });
    });
};

/*
 *
 * */
var confirmSendMsg = function () {

};


/*
* util
* */
function isEmptyObject(obj){
    for(var name in obj ){
        if(obj.hasOwnProperty(name)){
            return false;
        }
    }
    return true;
}

/**
 * insertCollection
 * @type {Function}
 * @param insertObj
 * {
        noticationuuid: '',
        datetime: '',
        title:'',
        content: '',
        noticationwriter:'',
        confirms:[{'':false,name:''},{'':false,name:''},{'':false,name:''}....]
    }
 */
var insertCollection = function (collectionName,insertObj,res,callback) {
    if( typeof insertObj === 'object' && typeof collectionName === 'string' ){
        MongoClient.connect(mongodbURL, function(err, db) {
            if(err) throw err;

            var thisCollection = db.collection(collectionName);
            //回调：
                if(callback && typeof callback === 'function' ){

                    thisCollection.insertOne(insertObj, function (err, result) {
                        if(err) throw err;
                        //是否延迟
                        if(insertObj.sendtime && insertObj.sendtime.length !== 0){
                            //延迟shedule
                            var job = {
                                noticationuuid:insertObj.noticationuuid,
                                sendtime:insertObj.sendtime
                            };
                            callback(job);
                        }else{
                            //即时发送
                            //将uuid插入每个user下的unconfirm字段
                            var userCollection = db.collection('user');
                            for(var temp in insertObj.confirms){
                                var filter = {'userid':temp},
                                    updater = {$push : {'notication.unconfirmed':insertObj.noticationuuid}};
                                userCollection.updateMany(filter,updater, function (err,result) {
                                    if(err) throw err;
                                    if(result && result.result.n > 0){}else{
                                        console.log('推送消息已更新到user下失败！');
                                    };
                                });
                                //即时 推送
                                //console.log('keyname: '+keyname);
                                callback(temp);
                            }
                        }

                        setTimeout(function () {
                            db.close();
                        },100);
                    });
                }
                res.send(JSON.stringify({'uuid':insertObj.noticationuuid}));
        });
    }
};

/**
 * 通过userid获取未读推送消息
 * @param userid
 * @param res
 */
var getUnreadsByUserid = function (userid, res) {
    MongoClient.connect(mongodbURL, function(err, db) {
        if(err) throw err;

        var userCollection = db.collection('user');
        var filter = {'userid' : userid};
        userCollection.find(filter).toArray(function (err, item) {
            //res.send(item[0].notication.unconfirmed);
            var sendnotications = [];
            //查询notication表
            if( item && item.length === 1 ){
                var noticationCollection = db.collection('notication');

                for( var i in item[0].notication.unconfirmed ){
                    var filter_notication = {'noticationuuid': item[0].notication.unconfirmed[i]};
                    noticationCollection.find(filter_notication).toArray(function (err, item) {
                        if( item && item.length === 1){
                            var temp_notication = {
                                noticationuuid: item[0].noticationuuid,
                                title: item[0].title,
                                noticationContent: item[0].content
                            };
                            sendnotications.push(temp_notication);
                        }
                    });
                }
            }

            setTimeout(function () {
                db.close();
                res.send(sendnotications);
            },100);
        });
    });
};

/**
 * 更新 未读消息 至  已读消息
 * @param confirmdata
 * {userid:'',uuid:''}
 * @param res
 */
var updateUnreadByuserid = function (confirmdata, res,rtc) {
    if( typeof confirmdata === 'object' ){
        MongoClient.connect(mongodbURL, function(err, db) {
                if(err) throw err;

                var userCollection = db.collection('user');

                //删除存在的未读消息
                var filter = {userid:confirmdata.userid};
                var updaterPull = {$pull:{'notication.unconfirmed':confirmdata.uuid}};
                userCollection.updateMany(filter,updaterPull, function (err, result) {
                    if(err) throw err;
                    //console.log(result);
                    if(result && result.result.n > 0){
                        var updaterPush = {$push:{'notication.confirmed':confirmdata.uuid}};
                        //加入已读消息
                        userCollection.updateMany(filter,updaterPush, function (err, _result) {
                            if(err) throw err;

                            //更新notication中的状态
                            var noticationCollection = db.collection('notication'),
                                filter = {noticationuuid:confirmdata.uuid},
                                updater = JSON.parse('{"$set":{"confirms.'+confirmdata.userid+'.isread":true}}');

                            noticationCollection.findOneAndUpdate(filter,updater,{}, function (err, result) {
                                if(err) throw err;
                                if(result && result.value){

                                    var noticationwriterid = result.value.noticationwriter,
                                        filter = {userid:noticationwriterid};

                                    userCollection.find(filter).limit(1).toArray(function (err, item) {
                                        if(err) throw err;

                                        //console.log(item);
                                        if( item && item.length === 1){
                                            var tempnoticationsocketid = item[0].tempnoticationsocketid,
                                                tempnoticationsocket = rtc.getSocket(tempnoticationsocketid);
                                            //如果存在临时socket，发送
                                            if(tempnoticationsocket){
                                                tempnoticationsocket.send(JSON.stringify({
                                                    "eventName": "onconfirm",
                                                    "data": {
                                                        "uuid": confirmdata.uuid,
                                                        "userid":confirmdata.userid
                                                    }
                                                }));
                                            }
                                        }

                                        setTimeout(function () {
                                            db.close();
                                        },100);

                                        res.send(200);
                                    });


                                }
                            });
                        });
                    }
                });
            }
        );
    }
};

/**
 * 获取已读人员
 * @param uuid
 * @param res
 */
var findreadByuuid = function (uuid,res) {
    MongoClient.connect(mongodbURL, function(err, db) {

        var noticationCollection = db.collection('notication');

        var filter = {noticationuuid:uuid};
        noticationCollection.find(filter).limit(1).next(function (err, item) {
            if(err) throw err;
            //console.log(item.confirms);
            var resJson = [],
                keyname;
            for(keyname in item.confirms){
                if(item.confirms[keyname].isread){
                    resJson.push(keyname);
                }
            }
            db.close();
            res.send(JSON.stringify(resJson));
        });
    });
};

/**
 * 通过uuid获取推送详情
 * @param uuid
 */
var getNoticationByuuid = function (uuid,callback) {
    MongoClient.connect(mongodbURL, function(err, db) {
        if(err) throw err;

        var collectionNotication = db.collection('notication'),
            filter = {noticationuuid:uuid};

        collectionNotication.find(filter).limit(1).next(function (err, item) {
            if(err) throw err;

            callback(item,db);
        })
    });
};

/**
 * 获取用户的推送消息历史纪录
 * @param userid
 * @param res
 */
var getNoticationHistoryByuserid = function (userid, res) {
    MongoClient.connect(mongodbURL, function(err, db) {
        if(err) throw err;

        var collectionNotication = db.collection('notication'),
            filter = {noticationwriter:userid},
            sorter = {datetime:-1};//降序 ; 升序值为1

        collectionNotication.find(filter).sort(sorter).toArray(function (err, docs) {
            if(err) throw err;

            //返回uuid,创建时间,标题
            var response=[],
                i;

            for(i=0;i<docs.length;i++){

                var datetime = new Date(parseInt(docs[i].datetime)).toLocaleString().replace(',','');

                var temp = {
                    noticationuuid:docs[i].noticationuuid,
                    datetime:datetime,
                    title:docs[i].title
                };

                response.push(temp);
            }

            res.send(response);
            db.close();
        });
    });
};

/**
 * 对应临时notication socket id
 * @param comingdata
 * @param res
 */
var setTempNoticationSocketId = function (comingdata, res) {
    MongoClient.connect(mongodbURL, function(err, db) {
        if(err) throw err;

        var userNotication = db.collection('user'),
            filter = {userid:comingdata.userid},
            updater = {$set:{tempnoticationsocketid:comingdata.tempnoticationsocketid}};


        userNotication.findOneAndUpdate(filter,updater,{},function (err, result) {
            if(err) throw err;

            //console.log(result);
            var response = {result : false};
            if(result.value){
                response.result = true
            }
            res.send(response);
            db.close();
        });
    });
};

module.exports.validLogin = validLogin;
module.exports.mapTempSocketid = mapTempSocketid;
module.exports.clearTempSocketid = clearTempSocketid;
module.exports.getUseridBySocketid = getUseridBySocketid;
module.exports.recordSendMsg = recordSendMsg;
module.exports.confirmSendMsg = confirmSendMsg;
module.exports.insertCollection = insertCollection;
module.exports.getUnreadsByUserid = getUnreadsByUserid;
module.exports.updateUnreadByuserid = updateUnreadByuserid;
module.exports.findreadByuuid = findreadByuuid;
module.exports.getNoticationByuuid = getNoticationByuuid;
module.exports.getNoticationHistoryByuserid = getNoticationHistoryByuserid;
module.exports.setTempNoticationSocketId = setTempNoticationSocketId;
