// notication/index.js
/**
 * Created by mac on 15/10/17.
 */
'use strict';
var uuid = require('node-uuid'),
    httputil = require('httpUtil'),
    egoomongodb = require('egoomongodb'),
    schedule = require('node-schedule'),
    JOBS;
//  "groupurl": "http://221.6.31.42:28080/SSIL/group.do?userid=jiangsudiaodu&json=1",
//  "messageurl":"http://221.6.31.42:28080/s2sh/message/message_queryAll?param=all"
/**
 * 定时任务
 * @param date
 */
function ontimeNotication(date){
    //年、月份0-11、日、时、分、秒
    var date = new Date(2015,10,2,15,7,30);
    var j = schedule.scheduleJob(date, function(){
        console.log("执行任务");
        j.cancel();
    });
    //JOBS.push(j);
}
ontimeNotication();

function Notication(){

}
Notication.prototype.delivery = function (rtc, msg) {
    rtc.emit('notication',msg);
};

/**
 *
 * @param incomingdata
 * {
 *  title:'',
 *  content:'',
 *  confirms:['','','',......]
 * }
 * @param res
 */
var insertAndpost = function (incomingdata ,res) {
    var tempinsertobj = {
        noticationuuid: uuid.v4(),
        datetime: new Date().getTime(),
        title:incomingdata.title,
        content: incomingdata.content,
        confirms: function () {
                var _confirms = '';
                for(var i in incomingdata.confirms){
                    var tempstring = '{"'+incomingdata.confirms[i]+'":false},';
                    _confirms += tempstring;
                }
                _confirms = '['+_confirms.substring(0,_confirms.length-1)+']';
            //console.log(_confirms);
            return JSON.parse(_confirms);
        }() };
    egoomongodb.insertCollection('notication',tempinsertobj,res,httputil.post);
};

//
var getNoticationByUserid = function (userid,res) {
    egoomongodb.getUnreadsByUserid(userid,res);
};

/**
 * 确认已读推送消息
 * @param confirmdata
 * {userid:'',uuid:''}
 * @param res
 */
var confirmNotication = function (confirmdata, res) {
    egoomongodb.updateUnreadByuserid(confirmdata, res);
};

/**
 * 获取已读人员
 * @param uuid
 * @param res
 */
var getconfirmed = function (uuid, res) {
    egoomongodb.findreadByuuid(uuid,res);
};
/**
 * 获取群组
 * @param res
 */
var getgroup = function (res) {
    httputil.getgroup(res);
};
/**
 * 获取短息模版
 * @param res
 */
var getmessage = function (res) {
    httputil.getmessage(res);
};

module.exports.notication = new Notication();
module.exports.insertAndpost = insertAndpost;
module.exports.getNoticationByUserid = getNoticationByUserid;
module.exports.confirmNotication = confirmNotication;
module.exports.getconfirmed = getconfirmed;
module.exports.getgroup = getgroup;
module.exports.getmessage = getmessage;
