// httpUtil/index.js
/**
 * Created by mac on 15/10/19.
 */
var http = require('http'),
    req = require('request'),
    querystring = require('querystring'),
    egoomongodb = require('egoomongodb'),
    package = require('../../package.json'),
    notificationurl = package.notificationurl,
    groupurl = package.groupurl,
    messageurl = package.messageurl;

/**
 *
 * @param insertobj
 * {
        noticationuuid: '',
        datetime: '',
        title:'',
        content: '',
        confirms:[{'':false},{'':false},{'':false}....]
    }
 */
var postNitification = function (recvid) {
    //data
    var reqData = querystring.stringify({
        sendid:'pushmessager',//统一发送者
        recvid:recvid,
        msgbody: ''
    });
    //post option
    var thishost = notificationurl.split(':')[0],
        thisport = notificationurl.split(':')[1];
    console.log(thishost);
    console.log(thisport);
    var opt = {
        method: "POST",
        host: thishost,
        port: thisport,
        path: "/pushmsg.do",
        headers: {
            "Content-Type": 'application/x-www-form-urlencoded;charset=UTF-8',
            "Content-Length": reqData.length
        }
    };
    //http
    var post_req = http.request(opt,function (response) {

        var responseText=[];
        var size = 0;
        response.on('data', function (data) {
            if(data === 'true'){
                console.log('BODY: ' + data);
            }
        });
        response.on('end', function () {
            console.log('发送成功！');
        });

    }).on('error', function (err) {
        console.log(err);
    });

    post_req.write(reqData);
    post_req.end();

};

//post('jiangsudiaodu');



module.exports.postNitification = postNitification;

function requestfunc (url,callback){
    //http
    var post_req = req(url,function (error, response, body) {
        if (!error && response.statusCode == 200) {
            var obj = JSON.parse(body);
            callback(obj);
            //console.log(obj);
        }
    });
}

/**
 * 获取群组接口
 * @param res
 */
var getgroup  = function (res) {
    requestfunc(groupurl, function (obj) {
        res.send(obj);
    });
};
module.exports.getgroup = getgroup;
//getgroup();

/**
 * 获取群组接口
 * @param res
 */
var getmessage  = function (res) {
    requestfunc(messageurl, function (obj) {
        res.send(obj);
    });
};
module.exports.getmessage = getmessage;
