// httpUtil/index.js
/**
 * Created by mac on 15/10/19.
 */
var http = require('http'),
    req = require('request'),
    querystring = require('querystring'),
    egoomongodb = require('egoomongodb'),
    package = require('../../package.json'),
    notificationurl = package.notificationurl,
    groupurl = package.groupurl,
    messageurl = package.messageurl,
    smsurl = package.smsurl;

/**
 *
 * @param insertobj
 * {
        noticationuuid: '',
        datetime: '',
        title:'',
        content: '',
        confirms:[{'':false},{'':false},{'':false}....]
    }
 */
var postNitification = function (recvid) {
    //data
    var reqData = querystring.stringify({
        sendid:'pushmessager',//统一发送者
        recvid:recvid,
        msgbody: '123'
    });
    //post option
    var thishost = notificationurl.split(':')[0],
        thisport = notificationurl.split(':')[1];
    var opt = {
        method: "POST",
        host: thishost,
        port: thisport,
        path: "/pushmsg.do",
        headers: {
            "Content-Type": 'application/x-www-form-urlencoded;charset=UTF-8',
            "Content-Length": reqData.length
        }
    };
    //http
    var post_req = http.request(opt,function (response) {

        var responseText=[];
        var size = 0;
        response.on('data', function (data) {
            if(data === 'true'){
                console.log('BODY: ' + data);
            }
        });
        response.on('end', function () {
            console.log('发送成功！');
        });

    }).on('error', function (err) {
        console.log(err);
    });

    post_req.write(reqData);
    post_req.end();

};

//post('jiangsudiaodu');
module.exports.postNitification = postNitification;

function requestfunc (url,callback){
    //http
    var post_req = req(url,function (error, response, body) {
        if (!error && response.statusCode == 200) {
            var obj = JSON.parse(body);
            callback(obj);
            //console.log(obj);
        }
    });
}

/**
 * 获取群组接口
 * @param res
 */
var getgroup  = function (userid,res) {
    var _groupurl = groupurl+userid;
    requestfunc(_groupurl, function (obj) {
        res.send(obj);
    });
};
module.exports.getgroup = getgroup;
//getgroup();

/**
 * 获取群组接口
 * @param res
 */
var getmessage  = function (res) {
    requestfunc(messageurl, function (obj) {
        res.send(obj);
    });
};
module.exports.getmessage = getmessage;


/**
 * 发送短信接口
 * @param res
 */
var sendsms  = function (nums,content,callback) {
	//data
    var reqData = querystring.stringify({
        nums:nums,//统一发送者
        contents:content
    });
    //post option
    var thishost = notificationurl.split(':')[0],
        thisport = notificationurl.split(':')[1];
    var opt = {
        method: "POST",
        host: thishost,
        port: thisport,
        path: "/easysendsms.do",
        headers: {
            "Content-Type": 'application/x-www-form-urlencoded;charset=UTF-8',
            "Content-Length": reqData.length
        }
    };
    //http
    var post_req = http.request(opt,function (response) {

        var responseText=[];
        var size = 0;
        response.on('data', function (data) {   
                console.log('BODY: ' + data);
        });
        response.on('end', function () {
            console.log('发送成功！');
        });

    }).on('error', function (err) {
        console.log(err);
    });

    post_req.write(reqData);
    post_req.end();
	
	
};
module.exports.sendsms = sendsms;


////test
//var posttest = function () {
//    //data
//    var reqData = querystring.stringify({
//        userid:'yanghong',//统一发送者
//        groupid:'huiyiceshi-tianjindiaodu',
//        docsharename: '大家好'
//    });
//    //post opti
//    var opt = {
//        method: "POST",
//        host: '221.6.31.42',
//        port: '28080',
//        path: "/SSIL/docname.do",
//        headers: {
//            "Content-Type": 'application/x-www-form-urlencoded;charset=UTF-8',
//            "Content-Length": reqData.length
//        }
//    };
//    //http
//    var post_req = http.request(opt,function (response) {
//
//        var responseText=[];
//        var size = 0;
//        response.on('data', function (data) {
//            console.log('BODY: ' + data);
//        });
//        response.on('end', function () {
//            console.log('发送成功！');
//        });
//
//    }).on('error', function (err) {
//        console.log(err);
//    });
//
//    post_req.write(reqData);
//    post_req.end();
//
//};
//posttest();
