// httpUtil/index.js
/**
 * Created by mac on 15/10/19.
 */
var http = require('http'),
    req = require('request'),
    querystring = require('querystring'),
    egoomongodb = require('egoomongodb');

/**
 *
 * @param insertobj
 * {
        noticationuuid: '',
        datetime: '',
        title:'',
        content: '',
        confirms:[{'':false},{'':false},{'':false}....]
    }
 */
var post = function (recvid) {
    //msgbody
    //var notications = JSON.stringify([
    //    {
    //        noticationuuid:'uuid1',
    //        title:'this is the title1',
    //        noticationContent: 'this is a content1 !'
    //    }
    //]);
    //data
    var reqData = querystring.stringify({
        sendid:'pushmessager',//统一发送者
        recvid:recvid,
        msgbody: ''
    });
    //console.log(reqData);
    //post option
    var opt = {
        method: "POST",
        host: "221.6.31.42",
        port: '8099',
        path: "/pushmsg.do",
        headers: {
            "Content-Type": 'application/x-www-form-urlencoded;charset=UTF-8',
            "Content-Length": reqData.length
        }
    };
    //http
    var post_req = http.request(opt,function (response) {

        var responseText=[];
        var size = 0;
        response.on('data', function (data) {
            if(data === 'true'){
                console.log('BODY: ' + data);
            }
        });
        response.on('end', function () {
            //// Buffer 是node.js 自带的库，直接使用
            //responseText = Buffer.concat(responseText,size);
            console.log('发送成功！');
        });

    }).on('error', function (err) {
        console.log(err);
    });

    post_req.write(reqData);
    post_req.end();

};

//post();



module.exports.post = post;



var test  = function () {

    var geturl = "http://221.6.31.42:28080/SSIL/group.do?userid=jiangsudiaodu&json=1";

    //var reqData = JSON.stringify({
    //    userid:'jiangsudiaodu',
    //    json: 1
    //});
    //var opt = {
    //    method: "POST",
    //    host: "http://221.6.31.42",
    //    port: '28080',
    //    path: "/SSIL/group.do",
    //    headers: {
    //        "Content-Type": 'application/x-www-form-urlencoded;charset=UTF-8',
    //        "Content-Length": reqData.length
    //    }
    //};
    //http
    var post_req = req(geturl,function (error, response, body) {
        if (!error && response.statusCode == 200) {
            var obj = JSON.parse(body);
            console.log(obj);
        }
    });
};

test();
